[TOPIC-BLE-LLCP] Add CTE Request control procedure

ppryga-nordic opened the pull request:
PR adds CTE Request procedure.

The procedure is required for implementation of Direction Finding in connected mode.

The PR provides implementation of the procedure in refactored LLCPs.
There are missing:
- integration with HCI
- LLL handling of CTE request, response PDUs
- periodicity of the procedure (host may requested controller to run the procedure in intervals).

ppryga-nordic committed:
Bluetooth: controller: Add LL_CTE_REQ pdu type

Add PDU structure for LL_CTE_REQ.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add LL_CTE_RSP pdu type

Add PDU structure for LL_CTE_RSP.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: LLCP: Add helpers for LL_CTE_REQ

Add PDU encode and verify helpers for LL_CTE_REQ.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: LLCP: Enable helper feats. of DF

Enable features related with direction finding:
- CTE request
- CTE response
- Antenna switch during CTE transmission (AoD)
- Antenna switch during CTE receive (AoA)
- CTE receive and sampling
Update naming of features related with DF to align
with names used in ll_feat.h.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: LLCP: Add Direction Finding Kconfig settings

Add Kconfiguration settings required to run direction
finding implementation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add structs for CTE req and rsp configuration

Constant Tone Extension request procedure requires two types of
configuration for:
- local proce CTE receive and sampling parameters,
- remote procedure CTE transmis parameters.

There are created two new structures to store those configurations.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLCP: Add API to enable or disable CTE req and rsp

Add functions that allow:
- start CTE request controle procedure.
- enable possibility to response for CTE request procedures.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLCP: Add encode decode func for CTE req and rsp PDU

Add helper functions that provide encode and decode functionality
for CTE: reuqests, responses and ULL notification PDUs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLCP: Add helpers to encode, decode reject ind

Add functions that encode and decode rejec extended
inditcation PDUs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLCP: Add handling of CTE req in common FSM

Add handling of CTE request control procedure in
the common code of refactored control procedures finite
state machine.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add handling of CTE req in local procedures

Add handling of CTE request control procedure  in local control
procedures.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add handling of CTE req in remote procedures

Add handling of CTE request control procedure  in remote control
procedures.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: LLCP: Add PDU helpers for CTE request proc

Add helpers responsible for:
- encoding CTE request pdu_data and node_rx_pdu
  for CTE request,
- verification of received PDUs for CTE requests.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: LLCP: Add unit tests for CTE request control proc

Add unit tests for CTE request control procedure.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add struct for IQ samples rep in conn mode

Add data type to store data related with Direction Finding IQ
samples report in connected mode.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

erbr-ot commented on tests/bluetooth/controller/common/include/helper_features.h:
```diff
@@ -111,46 +111,46 @@
 
 #if defined(CONFIG_MISSING)
-#define FEAT_CTE_TX 0x80000
+#define FEAT_CONNECTIONLESS_CTE_TX 0x80000
```
Is it on purpose that this #define is still under #if defined(CONFIG_MISSING)?

ppryga replied:
Yes, this feature is irrelevant for LLCP. It is a feature of CTE that works only with periodic advertising. There is no requirement to enable it for LLCP tests, similar to `FEAT_PER_ADV_SYNC_TX` and `FEAT_PER_ADV_SYNC_RX`

erbr-ot commented on tests/bluetooth/controller/common/include/helper_features.h:
```diff
@@ -111,46 +111,46 @@
 
 #if defined(CONFIG_MISSING)
-#define FEAT_CTE_RX 0x100000
+#define FEAT_CONNECTIONLESS_CTE_RX 0x100000
```
same as above

ppryga replied:
Answer is the same as above.

erbr-ot commented on tests/bluetooth/controller/ctrl_cte_req/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+common:
+    tags: test_framework bluetooth bt_le_cte_req bt_ull_llcp
+tests:
```
Should there be tests differentiating on the misc. configs (CONFIG_BT_CTLR_DF_CONN_CTE_REQ/RSP)?

ppryga replied:
That was may aim. To split tests into two separate parts: `CTE_REQ` and `CTE_RSP`. There are different configuration for request and response parts. It would be far more readable  if one part of the communication is removed from unit tests of another. The better approach is to concentrate on tests e.g. of `CTE_REQ` and mock peripheral part with expected responses.

erbr-ot reviewed (changes requested):
A first read through - generally it looks good to me

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -196,6 +202,29 @@ static void lp_comm_ntf_length_change(struct ll_conn *conn, struct proc_ctx *ctx
 }
 #endif /* CONFIG_BT_CTLR_DATA_LENGTH */
 
+static void lp_comm_ntf_cte_req(struct ll_conn *conn, struct proc_ctx *ctx, struct pdu_data *pdu)
```
If I understand you correctly Erik, you mean that the #if defined should be around the complete lp_comm_ntf_cte_req function, and then there should be an ifdef around line 253-255

erbr-ot replied:
Yeah, and that it seems that the #if defined() is not taken all the way, so to speak?

ppryga replied:
`#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_REQ)` moved to sorround whole function.

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:
```diff
@@ -1020,7 +1031,17 @@ static inline void pdu_encode_reject_ext_ind(struct pdu_data *pdu, uint8_t rejec
 	return ull_cp_priv_pdu_encode_reject_ext_ind(pdu, reject_opcode, error_code);
 }
 
+void ull_cp_priv_pdu_decode_reject_ext_ind(struct proc_ctx *ctx, struct pdu_data *pdu);
```
See PR 35867 where we replace the ull_cp_priv prefix with llcp_
Up to you if we do the same change in this PR or in a separate one

ppryga replied:
Changed to use prefix `llcp_`.

kruithofa commented on tests/bluetooth/controller/common/src/helper_util.c:
```diff
@@ -130,13 +134,43 @@ helper_pdu_ntf_verify_func_t *const helper_pdu_ntf_verify[] = {
 	[LL_LENGTH_REQ]            = NULL,
 	[LL_LENGTH_RSP]            = NULL,
+	[LL_CTE_REQ]               = NULL,
+	/* TODO (ppryga): Add verifiaction for RSP notification */
```
nitpick spelling: verifiACtion -> verification

ppryga replied:
Corrected, thank you for spotting.

ppryga commented:
The PR was just rebased on current topic-ble-llcp  branch. No comments were addressed.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp_pdu.c:
```diff
@@ -661,3 +681,47 @@ void ull_cp_priv_pdu_decode_length_rsp(struct ll_conn *conn,
+
+	p = &pdu->llctrl.cte_req;
+	p->min_cte_len_req = ctx->data.cte_req.min_len;
+	p->rfu = 0;
```
I've added comment, though the name of struct member seems to me to be self-explanatory.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -220,6 +250,9 @@ static void lp_comm_ntf(struct ll_conn *conn, struct proc_ctx *ctx)
 		lp_comm_ntf_length_change(conn, ctx, pdu);
 		break;
 #endif /* CONFIG_BT_CTLR_DATA_LENGTH */
+	case PROC_CTE_REQ:
```
Added `#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_REQ)`

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -310,6 +343,15 @@ static void lp_comm_complete(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
 		}
 		break;
 #endif /* CONFIG_BT_CTLR_DATA_LENGTH */
+	case PROC_CTE_REQ:
```
Added `#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_REQ)`

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -393,7 +435,14 @@ static void lp_comm_send_req(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
 		}
-
+		break;
+	case PROC_CTE_REQ:
```
Added `#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_REQ)`

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -804,6 +878,21 @@ static void rp_comm_send_rsp(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
 			ctx->state = RP_COMMON_STATE_WAIT_TX_ACK;
 		}
 		break;
+	case PROC_CTE_REQ:
```
Added `#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_REQ)`

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -804,6 +878,21 @@ static void rp_comm_send_rsp(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
 		}
 		break;
+	case PROC_CTE_REQ:
+		/* Check if LL_CTE_RSP is enabled and requested CTE parameters are valid for
```
The comment and the one below (`TODO (ppryga): Add parameters validation here...`) are not relevant now. As you have spot the params are checked in `rp_comm_tx()`. Comments were removed.

ppryga commented:
Resolved all comments. 
Modified `rp_comm_tx` to correctly handle CTE_REQ parameters (the parameters are verified in order from BT Core spec. 5.2 Vol. 6, Part B, 5.1.12 Constant Tone Extension Request procedure).

ppryga-nordic committed:
Bluetooth: controller: LLCP: fix warning defined but unused function

Fix warning: defined but unused function pu_select_phy_timing_restrict.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: LLCP: fix error in features exchange procedure

There was an error in the filter_feat function. There was missing
filtering of bits for features that are not valid between controllers.
That caused failures of unit tests when FEAT_ANT_SWITCH_CTE_TX and
FEAT_ANT_SWITCH_CTE_RX were added. These features are not valid
between controllers.
According to BT core specification Vol 6, PART B, section 4.46,
if feature bit is not valid between controllers is should be ignored.
Due to that, these bits are filtered out (set to zeros).

That also forced to update unit tests due to muiltiple failures.
There were wrong values for expedted feature sets.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Changes added:
- fixed build errors
- fixed issue in CTE reguest code that caused failing CTE request unit tests to fail,
- fixed issue in feature exchange procedure (for details see commit log: https://github.com/zephyrproject-rtos/zephyr/pull/35749/commits/7736b03b668286245b85636aceddb07f45c69633),
- updated tests for feature exchange procedure

ppryga-nordic committed:
Bluetooth: controller: Add missing #include "lll_conn.h"

There was a compilation error about missing declaration of
struct node_tx.
Add missing #include "lll_conn.h" to ull_adv_aux.c to fix that.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_local.c:
```diff
@@ -151,6 +151,9 @@ void llcp_lr_rx(struct ll_conn *conn, struct proc_ctx *ctx, struct node_rx_pdu *
 	case PROC_DATA_LENGTH_UPDATE:
 		llcp_lp_comm_rx(conn, ctx, rx);
 		break;
+	case PROC_CTE_REQ:
```
the case is within CONFIG_BT_CTRL_DATA_LENGTH define?
Should be under CTE_REQ define?

ppryga replied:
No, that is wrong. Corrected and added #define guards.

erbr-ot reviewed (changes requested):
Looks good, except for the few missing #ifdef's

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:
```diff
@@ -187,6 +188,12 @@ struct proc_ctx {
 			uint8_t chm[5];
 		} chmu;
 
+		/* Use by CTE Request Procedure */
```
Added missing #define guards.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp.h:
```diff
@@ -146,3 +146,13 @@ const uint8_t *ull_cp_chan_map_update_pending(struct ll_conn *conn);
  */
 uint8_t ull_cp_data_length_update(struct ll_conn *conn, uint16_t max_tx_octets, uint16_t max_tx_time);
 #endif /* CONFIG_BT_CTLR_DATA_LENGTH */
+/**
```
In case of declarations in header files conditional compilation is not needed. If the function is not used compiler will not complain about it. This is a common way for function declarations in header files in controller code.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -291,6 +294,9 @@ struct proc_ctx *llcp_create_remote_procedure(enum llcp_proc proc)
 		llcp_rp_comm_init_proc(ctx);
 		break;
 #endif /* CONFIG_BT_CTLR_DATA_LENGTH */
+	case PROC_CTE_REQ:
```
Added missing #define guards.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -228,6 +228,9 @@ struct proc_ctx *llcp_create_local_procedure(enum llcp_proc proc)
 		llcp_lp_comm_init_proc(ctx);
 		break;
 #endif /* CONFIG_BT_CTLR_DATA_LENGTH */
+	case PROC_CTE_REQ:
```
Added missing #define guards.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_llcp_local.c:
```diff
@@ -237,6 +240,10 @@ static void lr_act_run(struct ll_conn *conn)
 	case PROC_DATA_LENGTH_UPDATE:
 		llcp_lp_comm_run(conn, ctx, NULL);
 		break;
+	case PROC_CTE_REQ:
```
No, that is wrong. Corrected and added #define guards.

ppryga commented:
@erbr-ot I've corrected code according to most of your requests. Please take a look on the PR again.

erbr-ot reviewed (approved):
LGTM

ppryga commented:
Branch rebased onto latest topic-ble-llcp branch. 

@kruithofa is there anything else you would like to be changed?

ppryga-nordic committed:
Bluetooth: controller: Add missing #include "lll_df_types.h"

Part of the code was not building when DF is enabled.
There was missing include of lll_df_types.h.
In some places the include was just moved up, before
include of lll_conn.h file.

The lll_df_types.h requires util.h to be included before.

The commit adds missing includes or reordest available
includes.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
The PR was rebased on topic-ble-llcp again. There was a build error related with missing `#include "lll_df_types.h".
I've added new commit that fixes the problem.

@kruithofa @thoh-ot  @sjanc could you take a look? I hope it is merged soon because it has been waiting for quite a long time now.
